### PR TITLE
Use Groupie Kotlin ViewHolder and Item

### PR DESCRIPTION
**Pull request by veyndan**
A rework of #1178.

### CODE REVIEW

File: app/src/main/java/com/github/pockethub/android/ui/item/GitHubCommentItem.kt [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import com.github.pockethub.android.ui.comment.EditCommentListener
+import com.github.pockethub.android.util.AvatarLoader
+import com.github.pockethub.android.util.HttpImageGetter
+import com.github.pockethub.android.util.TimeUtils
+import com.meisolsson.githubsdk.model.GitHubComment
+import com.xwray.groupie.kotlinandroidextensions.Item
+import com.xwray.groupie.kotlinandroidextensions.ViewHolder
+import kotlinx.android.synthetic.main.comment.*
+
+class GitHubCommentItem(private val avatarLoader: AvatarLoader, private val imageGetter: HttpImageGetter, private val editCommentListener: EditCommentListener?, private val deleteCommentListener: DeleteCommentListener?, private val username: String, private val canWrite: Boolean, val gitHubComment: GitHubComment) : Item(gitHubComment.id()!!) {
```
Comment (Meisolsson) [inline app/src/main/java/com/github/pockethub/android/ui/item/GitHubCommentItem.kt]: This should not be all on one line, too long

File: app/src/main/java/com/github/pockethub/android/ui/item/TextItem.kt [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package com.github.pockethub.android.ui.item
+
+import android.support.annotation.IdRes
+import android.support.annotation.LayoutRes
+import android.widget.TextView
+import com.xwray.groupie.kotlinandroidextensions.Item
+import com.xwray.groupie.kotlinandroidextensions.ViewHolder
+
+class TextItem(@param:LayoutRes private val layoutId: Int, @param:IdRes private val textViewId: Int, private val text: CharSequence) : Item(text.hashCode().toLong()) {
```
Comment (Meisolsson) [inline app/src/main/java/com/github/pockethub/android/ui/item/TextItem.kt]: Too long line

File: app/src/main/java/com/github/pockethub/android/ui/item/code/BlobItem.kt [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import android.content.Context
+import android.text.format.Formatter
+import com.github.pockethub.android.R
+import com.github.pockethub.android.core.code.FullTree
+import com.github.pockethub.android.util.ServiceUtils
+import com.xwray.groupie.kotlinandroidextensions.Item
+import com.xwray.groupie.kotlinandroidextensions.ViewHolder
+import kotlinx.android.synthetic.main.blob_item.*
+
+class BlobItem(private val context: Context, val file: FullTree.Entry, private val indented: Boolean) : Item(file.entry.sha()!!.hashCode().toLong()) {
```
Comment (Meisolsson) [inline app/src/main/java/com/github/pockethub/android/ui/item/code/BlobItem.kt]: Line to long

File: app/src/main/java/com/github/pockethub/android/ui/item/code/FolderItem.kt [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import android.content.Context
+import com.github.pockethub.android.R
+import com.github.pockethub.android.core.code.FullTree
+import com.github.pockethub.android.core.commit.CommitUtils
+import com.github.pockethub.android.util.ServiceUtils
+import com.xwray.groupie.kotlinandroidextensions.Item
+import com.xwray.groupie.kotlinandroidextensions.ViewHolder
+import kotlinx.android.synthetic.main.folder_item.*
+
+class FolderItem(context: Context, val folder: FullTree.Folder, private val indented: Boolean) : Item(folder.entry.sha()!!.hashCode().toLong()) {
```
Comment (Meisolsson) [inline app/src/main/java/com/github/pockethub/android/ui/item/code/FolderItem.kt]: Line to long

File: app/src/main/java/com/github/pockethub/android/ui/item/commit/CommitCommentItem.kt [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+import com.github.pockethub.android.R
+import com.github.pockethub.android.util.AvatarLoader
+import com.github.pockethub.android.util.HttpImageGetter
+import com.github.pockethub.android.util.TimeUtils
+import com.meisolsson.githubsdk.model.git.GitComment
+import com.xwray.groupie.kotlinandroidextensions.Item
+import com.xwray.groupie.kotlinandroidextensions.ViewHolder
+import kotlinx.android.synthetic.main.comment.*
+
+class CommitCommentItem @JvmOverloads constructor(private val avatarLoader: AvatarLoader, private val imageGetter: HttpImageGetter, val comment: GitComment, private val isLineComment: Boolean = false) : Item(comment.id()!!) {
```
Comment (Meisolsson) [inline app/src/main/java/com/github/pockethub/android/ui/item/commit/CommitCommentItem.kt]: Line to long

File: app/src/main/java/com/github/pockethub/android/ui/item/commit/CommitHeaderItem.kt [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+import com.github.pockethub.android.R
+import com.github.pockethub.android.core.commit.CommitUtils
+import com.github.pockethub.android.ui.StyledText
+import com.github.pockethub.android.util.AvatarLoader
+import com.meisolsson.githubsdk.model.Commit
+import com.xwray.groupie.kotlinandroidextensions.Item
+import com.xwray.groupie.kotlinandroidextensions.ViewHolder
+import kotlinx.android.synthetic.main.commit_header.*
+
+class CommitHeaderItem(private val avatarLoader: AvatarLoader, private val context: Context, val commit: Commit) : Item(commit.sha()!!.hashCode().toLong()) {
```
Comment (Meisolsson) [inline app/src/main/java/com/github/pockethub/android/ui/item/commit/CommitHeaderItem.kt]: Line to long

File: app/src/main/java/com/github/pockethub/android/ui/item/issue/IssueEventItem.kt [hunk 7]
```diff
@@ -2,104 +2,75 @@ package com.github.pockethub.android.ui.item.issue
...
 import com.meisolsson.githubsdk.model.IssueEvent
 import com.meisolsson.githubsdk.model.IssueEventType.*
+import com.xwray.groupie.kotlinandroidextensions.Item
+import com.xwray.groupie.kotlinandroidextensions.ViewHolder
+import kotlinx.android.synthetic.main.comment_event.*
 import kotlinx.android.synthetic.main.comment_event.view.*
 
-class IssueEventItem(avatarLoader: AvatarLoader, private val context: Context, private val issue: Issue, dataItem: IssueEvent)
-    : BaseDataItem<IssueEvent, IssueEventItem.ViewHolder>(avatarLoader, dataItem, dataItem.id()!!.toLong()) {
+class IssueEventItem(private val avatarLoader: AvatarLoader, private val context: Context, private val issue: Issue, val issueEvent: IssueEvent) : Item(issueEvent.id()!!.toLong()) {
```
Comment (Meisolsson) [inline app/src/main/java/com/github/pockethub/android/ui/item/issue/IssueEventItem.kt]: Line to long

File: app/src/main/java/com/github/pockethub/android/ui/item/issue/IssueEventItem.kt [hunk 8]
```diff
@@ -2,104 +2,75 @@ package com.github.pockethub.android.ui.item.issue
...
+                holder.tv_event.setTextColor(context.resources.getColor(R.color.text_description))
             }
             Merged -> {
-                message += String.format(" commit <b>%s</b> into <tt>%s</tt> from <tt>%s</tt>", event.commitId()!!.substring(0, 6),
-                        issue.pullRequest()!!.base()!!.ref(),
-                        issue.pullRequest()!!.head()!!.ref())
-                holder.event.text = OcticonTextView.ICON_MERGE
-                holder.event.setTextColor(
-                        context.resources.getColor(R.color.issue_event_merged))
+                message += String.format(" commit <b>%s</b> into <tt>%s</tt> from <tt>%s</tt>", issueEvent.commitId()!!.substring(0, 6), issue.pullRequest()!!.base()!!.ref(), issue.pullRequest()!!.head()!!.ref())
```
Comment (Meisolsson) [inline app/src/main/java/com/github/pockethub/android/ui/item/issue/IssueEventItem.kt]: Line to long

File: app/src/main/java/com/github/pockethub/android/ui/item/issue/IssueFilterHeaderItem.kt [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+import android.view.View
+import com.github.pockethub.android.R
+import com.github.pockethub.android.core.issue.IssueFilter
+import com.github.pockethub.android.ui.issue.LabelDrawableSpan
+import com.github.pockethub.android.util.AvatarLoader
+import com.xwray.groupie.kotlinandroidextensions.Item
+import com.xwray.groupie.kotlinandroidextensions.ViewHolder
+import kotlinx.android.synthetic.main.issues_filter_details.*
+
+class IssueFilterHeaderItem(private val avatarLoader: AvatarLoader, val issueFilter: IssueFilter) : Item(issueFilter.hashCode().toLong()) {
```
Comment (Meisolsson) [inline app/src/main/java/com/github/pockethub/android/ui/item/issue/IssueFilterHeaderItem.kt]: Line to long

File: app/src/main/java/com/github/pockethub/android/ui/item/issue/IssueHeaderItem.kt [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+import com.github.pockethub.android.util.AvatarLoader
+import com.github.pockethub.android.util.HttpImageGetter
+import com.meisolsson.githubsdk.model.Issue
+import com.meisolsson.githubsdk.model.IssueState
+import com.xwray.groupie.kotlinandroidextensions.Item
+import com.xwray.groupie.kotlinandroidextensions.ViewHolder
+import kotlinx.android.synthetic.main.issue_header.*
+import kotlinx.android.synthetic.main.milestone.*
+
+class IssueHeaderItem(private val avatarLoader: AvatarLoader, private val imageGetter: HttpImageGetter, private val context: Context, private val actionListener: OnIssueHeaderActionListener, val issue: Issue) : Item(issue.id()!!) {
```
Comment (Meisolsson) [inline app/src/main/java/com/github/pockethub/android/ui/item/issue/IssueHeaderItem.kt]: Line too long

File: app/src/main/java/com/github/pockethub/android/ui/item/news/IssueCommentEventItem.kt [hunk 11]
```diff
@@ -9,38 +9,40 @@ import com.github.pockethub.android.util.AvatarLoader
 import com.meisolsson.githubsdk.model.GitHubComment
 import com.meisolsson.githubsdk.model.GitHubEvent
 import com.meisolsson.githubsdk.model.payload.IssueCommentPayload
+import com.xwray.groupie.kotlinandroidextensions.ViewHolder
+import kotlinx.android.synthetic.main.news_item.*
 
-class IssueCommentEventItem(avatarLoader: AvatarLoader, dataItem: GitHubEvent) : NewsItem(avatarLoader, dataItem) {
+class IssueCommentEventItem(avatarLoader: AvatarLoader, override val gitHubEvent: GitHubEvent) : NewsItem(avatarLoader, gitHubEvent) {
```
Comment (Meisolsson) [inline app/src/main/java/com/github/pockethub/android/ui/item/news/IssueCommentEventItem.kt]: Line is too long and it feels weird to override the `gitHubEvent` here

File: app/src/main/java/com/github/pockethub/android/ui/item/news/IssuesEventItem.kt [hunk 12]
```diff
@@ -7,25 +7,27 @@ import com.github.pockethub.android.ui.view.OcticonTextView
 import com.github.pockethub.android.util.AvatarLoader
 import com.meisolsson.githubsdk.model.GitHubEvent
 import com.meisolsson.githubsdk.model.payload.IssuesPayload
+import com.xwray.groupie.kotlinandroidextensions.ViewHolder
+import kotlinx.android.synthetic.main.news_item.*
 
-class IssuesEventItem(avatarLoader: AvatarLoader, dataItem: GitHubEvent) : NewsItem(avatarLoader, dataItem) {
+class IssuesEventItem(avatarLoader: AvatarLoader, override val gitHubEvent: GitHubEvent) : NewsItem(avatarLoader, gitHubEvent) {
```
Comment (Meisolsson) [inline app/src/main/java/com/github/pockethub/android/ui/item/news/IssuesEventItem.kt]: Overriding feel wrong here

File: app/src/main/java/com/github/pockethub/android/ui/item/news/MemberEventItem.kt [hunk 13]
```diff
@@ -6,23 +6,25 @@ import com.github.pockethub.android.ui.view.OcticonTextView
 import com.github.pockethub.android.util.AvatarLoader
 import com.meisolsson.githubsdk.model.GitHubEvent
 import com.meisolsson.githubsdk.model.payload.MemberPayload
+import com.xwray.groupie.kotlinandroidextensions.ViewHolder
+import kotlinx.android.synthetic.main.news_item.*
 
-class MemberEventItem(avatarLoader: AvatarLoader, dataItem: GitHubEvent) : NewsItem(avatarLoader, dataItem) {
+class MemberEventItem(avatarLoader: AvatarLoader, override val gitHubEvent: GitHubEvent) : NewsItem(avatarLoader, gitHubEvent) {
```
Comment (Meisolsson) [inline app/src/main/java/com/github/pockethub/android/ui/item/news/MemberEventItem.kt]: Overridning feel wrong

File: app/src/main/java/com/github/pockethub/android/ui/item/news/NewsItem.kt [hunk 14]
```diff
@@ -1,33 +1,29 @@
...
 import com.meisolsson.githubsdk.model.User
-import kotlinx.android.synthetic.main.news_item.view.*
+import com.xwray.groupie.kotlinandroidextensions.Item
+import com.xwray.groupie.kotlinandroidextensions.ViewHolder
+import kotlinx.android.synthetic.main.news_item.*
 
-open class NewsItem(avatarLoader: AvatarLoader, dataItem: GitHubEvent) :
-        BaseDataItem<GitHubEvent, NewsItem.ViewHolder>(avatarLoader, dataItem,
-                dataItem.id()!!.hashCode().toLong()) {
+open class NewsItem(private val avatarLoader: AvatarLoader, open val gitHubEvent: GitHubEvent) : Item(gitHubEvent.id()!!.hashCode().toLong()) {
```
Comment (Meisolsson) [inline app/src/main/java/com/github/pockethub/android/ui/item/news/NewsItem.kt]: The line is too long

File: app/src/main/java/com/github/pockethub/android/ui/item/news/PublicEventItem.kt [hunk 15]
```diff
@@ -5,19 +5,21 @@ import com.github.pockethub.android.ui.StyledText
 import com.github.pockethub.android.ui.view.OcticonTextView
 import com.github.pockethub.android.util.AvatarLoader
 import com.meisolsson.githubsdk.model.GitHubEvent
+import com.xwray.groupie.kotlinandroidextensions.ViewHolder
+import kotlinx.android.synthetic.main.news_item.*
 
-class PublicEventItem(avatarLoader: AvatarLoader, dataItem: GitHubEvent) : NewsItem(avatarLoader, dataItem) {
+class PublicEventItem(avatarLoader: AvatarLoader, override val gitHubEvent: GitHubEvent) : NewsItem(avatarLoader, gitHubEvent) {
```
Comment (Meisolsson) [inline app/src/main/java/com/github/pockethub/android/ui/item/news/PublicEventItem.kt]: Why are we overriding here

File: app/src/main/java/com/github/pockethub/android/ui/item/news/PullRequestReviewCommentEventItem.kt [hunk 16]
```diff
@@ -8,29 +8,29 @@ import com.github.pockethub.android.util.AvatarLoader
 import com.meisolsson.githubsdk.model.GitHubEvent
 import com.meisolsson.githubsdk.model.ReviewComment
 import com.meisolsson.githubsdk.model.payload.PullRequestReviewCommentPayload
+import com.xwray.groupie.kotlinandroidextensions.ViewHolder
+import kotlinx.android.synthetic.main.news_item.*
 
-class PullRequestReviewCommentEventItem(avatarLoader: AvatarLoader, dataItem: GitHubEvent) : NewsItem(avatarLoader, dataItem) {
+class PullRequestReviewCommentEventItem(avatarLoader: AvatarLoader, override val gitHubEvent: GitHubEvent) : NewsItem(avatarLoader, gitHubEvent) {
```
Comment (Meisolsson) [inline app/src/main/java/com/github/pockethub/android/ui/item/news/PullRequestReviewCommentEventItem.kt]: Line is too long

File: app/src/main/java/com/github/pockethub/android/ui/item/notification/NotificationHeaderItem.kt [hunk 17]
```diff
@@ -0,0 +1,26 @@
+package com.github.pockethub.android.ui.item.notification
+
+import com.github.pockethub.android.R
+import com.github.pockethub.android.ui.notification.NotificationListFragment
+import com.meisolsson.githubsdk.model.Repository
+import com.xwray.groupie.kotlinandroidextensions.Item
+import com.xwray.groupie.kotlinandroidextensions.ViewHolder
+import kotlinx.android.synthetic.main.notification_item_header.*
+
+class NotificationHeaderItem(val repository: Repository, private val notificationReadListener: NotificationListFragment) : Item(repository.id()!!) {
```
Comment (Meisolsson) [inline app/src/main/java/com/github/pockethub/android/ui/item/notification/NotificationHeaderItem.kt]: Line is too long

File: app/src/main/java/com/github/pockethub/android/ui/item/notification/NotificationItem.kt [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+import com.github.pockethub.android.R
+import com.github.pockethub.android.ui.notification.NotificationListFragment
+import com.github.pockethub.android.ui.view.OcticonTextView
+import com.github.pockethub.android.util.TimeUtils
+import com.meisolsson.githubsdk.model.NotificationThread
+import com.xwray.groupie.kotlinandroidextensions.Item
+import com.xwray.groupie.kotlinandroidextensions.ViewHolder
+import kotlinx.android.synthetic.main.notification_item.*
+
+class NotificationItem(val notificationThread: NotificationThread, private val notificationReadListener: NotificationListFragment) : Item(notificationThread.id()!!.hashCode().toLong()) {
```
Comment (Meisolsson) [inline app/src/main/java/com/github/pockethub/android/ui/item/notification/NotificationItem.kt]: Line is too long

File: app/src/main/java/com/github/pockethub/android/ui/item/notification/NotificationItem.kt [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+        }
+
+        holder.tv_notification_title.text = notificationThread.subject()!!.title()
+        holder.tv_notification_date.text = TimeUtils.getRelativeTime(notificationThread.updatedAt())
+
+        holder.tv_notification_read_icon.text = OcticonTextView.ICON_READ
+        holder.tv_notification_read_icon.visibility = if (notificationThread.unread()!!) View.VISIBLE else View.GONE
+
+        if (!holder.tv_notification_read_icon.hasOnClickListeners()) {
+            holder.tv_notification_read_icon.setOnClickListener { _ -> notificationReadListener.readNotification(notificationThread) }
```
Comment (Meisolsson) [inline app/src/main/java/com/github/pockethub/android/ui/item/notification/NotificationItem.kt]: Line is too long

**Comment by veyndan** [comment]
I removed the overrides and formatted the code. The link you referenced didn't talk about formatting class signatures, so I used the [official Kotlin style guide](https://kotlinlang.org/docs/reference/coding-conventions.html#class-header-formatting) instead.

**Pull request merged**